feat: add `stats/base/dists/planck/pmf`

Jaysukh-409 opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   Adds the `pmf` for the Planck Discrete Random Variable.

## Related Issues

> Does this pull request have any related issues?

This pull request:

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Jaysukh-409 committed:
feat: add stats/base/dists/planck

Jaysukh-409 commented:
I searched for the `Planck Distribution` in the Julia, but I wasn't able to find much. And due to that currently test seems to be incomplete. 
Should we use another language for the same task ? Like Python using SciPy to generate tests to check accuracy.

Jaysukh-409 committed:
test: fix the invalid tests

Planeshifter commented:
@Jaysukh-409 Yes, SciPy is a good reference and we should test against it when the distribution is not available in R or Julia.

Jaysukh-409 committed:
test: add test for large and small shape parameter

Jaysukh-409 commented:
@Planeshifter I tried to test against the scipy but some of my tests are failing. Is there anything which I missed in implementation?

Jaysukh-409 committed:
fix: change implementation to clear all test

Jaysukh-409 committed:
fix: lint errors

Jaysukh-409 commented:
@kgryte @Planeshifter PR is ready for review.

Jaysukh-409 committed:
fix: change type of lambda

Jaysukh-409 committed:
docs: update the condition description

Signed-off-by: Jaysukh Makvana <111515433+Jaysukh-409@users.noreply.github.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/planck/pmf/README.md:
```diff
@@ -0,0 +1,144 @@
+
+<section class="intro">
+
+The [probability mass function][pmf] (PMF) for a planck random variable is defined as
```
As `Planck` is a proper noun, it should be capitalized in all descriptions here and elsewhere.

kgryte replied:
s/planck/Planck/

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/planck/pmf/README.md:
````diff
@@ -0,0 +1,144 @@
+<!-- eslint no-undef: "error" -->
+
+```javascript
+var randu = require( '@stdlib/random/base/randu' );
+var round = require( '@stdlib/math/base/special/round' );
````
Let's go ahead and generate an array of random numbers using `@stdlib/random/array/discrete-uniform`. This ensures consistency with some of the recent work we've been doing when updating various dists packages.

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/planck/pmf/benchmark/benchmark.js:
```diff
@@ -0,0 +1,81 @@
+
+	b.tic();
+	for ( i = 0; i < b.iterations; i++ ) {
+		x = round( randu()*40.0 );
```
Same thing. Pre-generate an array of random values, rather than individually within a loop.

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/planck/pmf/benchmark/benchmark.js:
```diff
@@ -0,0 +1,81 @@
+
+	b.tic();
+	for ( i = 0; i < b.iterations; i++ ) {
+		x = randu() * 40.0;
```
Same thing.

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/planck/pmf/examples/index.js:
```diff
@@ -0,0 +1,35 @@
+
+'use strict';
+
+var randu = require( '@stdlib/random/base/randu' );
```
Same comment as README.

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/planck/pmf/test/fixtures/python/REQUIRE:
```diff
@@ -0,0 +1,3 @@
+numpy 2.2.0
+scipy 1.14.1
+python 3.10.12
```
Missing trailing newline.

kgryte reviewed (changes requested):
Thanks for working on this @Jaysukh-409. Left a few minor niggles, but overall looking good.

Jaysukh-409 committed:
fix: apply code review suggestion

Jaysukh-409 commented:
@kgryte This PR is now ready to merge.

kgryte commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into planck

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/planck/pmf/README.md:
````diff
@@ -0,0 +1,145 @@
+
+```javascript
+var discreteUniform = require( '@stdlib/random/array/discrete-uniform' );
+var randu = require( '@stdlib/random/base/randu' );
````
@Jaysukh-409 Use `random/array/uniform` for generating `lambda`, as well. Meaning, move RV generation out of the loop.

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/planck/pmf/benchmark/benchmark.js:
```diff
@@ -0,0 +1,81 @@
+	var y;
+	var i;
+
+	x = discreteUniform( b.iterations, 0, 40 );
```
Don't use `b.iterations`. Just `100`. `b.iterations` could be enormous and you'll cause the process to run out of memory.

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/planck/pmf/benchmark/benchmark.js:
```diff
@@ -0,0 +1,81 @@
+	var y;
+	var i;
+
+	x = discreteUniform( b.iterations, 0, 40 );
```
Same comment.

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/planck/pmf/benchmark/benchmark.js:
```diff
@@ -0,0 +1,81 @@
+
+	b.tic();
+	for ( i = 0; i < b.iterations; i++ ) {
+		lambda = ( randu()*10.0 );
```
You need to move this out of the loop. Then use modulo arithmetic below (e.g., `i % x.length`, etc). As I mentioned elsewhere, see other recent changes to base stats dists packages where we've already been doing this work.

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/planck/pmf/benchmark/benchmark.js:
```diff
@@ -0,0 +1,81 @@
+
+	x = discreteUniform( b.iterations, 0, 40 );
+	lambda = 0.3;
+	mypmf = pmf.factory( lambda );
```
Just provide `0.3` directly. No need to have a separate `lambda` variable.

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/planck/pmf/examples/index.js:
```diff
@@ -0,0 +1,36 @@
+var y;
+var i;
+
+x = discreteUniform( 10, 0, 5 );
```
Same comments as README.

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/planck/pmf/lib/factory.js:
```diff
@@ -0,0 +1,77 @@
+/**
+* Returns a function for evaluating the probability mass function (PMF) for a Planck distribution with shape parameter `lambda`.
+*
+* @param {number} lambda - shape parameter
```
```suggestion
* @param {NonNegativeNumber} lambda - shape parameter
```

Yes, we handle `lambda < 0.0`, but the intent is that users should provide a nonnegative number.

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/planck/pmf/lib/main.js:
```diff
@@ -0,0 +1,76 @@
+* Evaluates the probability mass function (PMF) for a Planck distribution with shape parameter `lambda` at a value `x`.
+*
+* @param {number} x - input value
+* @param {number} lambda - shape parameter
```
```suggestion
* @param {NonNegativeNumber} lambda - shape parameter
```

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/planck/pmf/test/fixtures/python/REQUIRE:
```diff
@@ -0,0 +1,3 @@
+numpy 2.2.0
```
What is this file? This should be removed. `REQUIRE` is a Julia thing, not Python. When in doubt, consult our snippets: https://github.com/stdlib-js/stdlib/tree/develop/tools/snippets/test/fixtures

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/planck/pmf/test/fixtures/python/runner.py:
```diff
@@ -0,0 +1,83 @@
+# !/usr/bin/env python
```
This file does not match our snippet runner: https://github.com/stdlib-js/stdlib/blob/develop/tools/snippets/test/fixtures/python/runner.py.

kgryte reviewed (changes requested):
Left another round of comments.

kgryte commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into planck

Jaysukh-409 committed:
fix: apply code review suggestion

Jaysukh-409 committed:
docs: update the function argument name

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Jaysukh-409 committed:
chore: minor clean up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: passed
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

Jaysukh-409 committed:
chore: clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Jaysukh-409 committed:
chore: update docs

Signed-off-by: Jaysukh Makvana <111515433+Jaysukh-409@users.noreply.github.com>

kgryte committed:
chore: clean-up and remove namespace files

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: missing_dependencies
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte reviewed (approved):
LGTM. @Jaysukh-409 I made a few small changes. Would be worth checking whether they are applicable to your other PRs for `cdf`, `logcdf`, etc. Cheers!
